### ISSUE TITLE
Re-enable corefx multifile test configurations once nutc error is fixed

### ISSUE THREAD

**Issue by joperezr** [dotnet/corefx#22826]
A recent change in PN side has broken the build for the uapaot configurations that use the multifile test configuration. In order to unblock the full official build, we turned off the multifile test configurations from corefx/master, but we should re-enable them once the fix has been checked in to PN.

cc: @joshfree @trylek

**Comment by joshfree** [comment]
Related change that will need to be undone once the PN side is fixed 7c62f4e

**Comment by ViktorHofer** [comment]
Is that issue still happening or can we already revert the change?

**Comment by joperezr** [comment]
That is a good question. I would try enabling them and see if the test run succeeds.

**Comment by wtgodbe** [comment]
@ViktorHofer @joshfree @ericstj @joperezr does anybody know where these packages live? I'm having trouble figuring out how we used to restore them

**Comment by trylek** [comment]
The nightly drops get published to the share

<pre>
\\cpvsbuild\drops\dev14\ProjectN\layouts\x86ret\27427.00\enu\netfx\projectnx86\net\netfx_ProjectNx86.msi
</pre>

(there's also the raw TestILC folder under that path). As these are internal only, I'm not sure if we publish them anywhere, most certainly not on any publicly available nuget feed :-).

**Comment by wtgodbe** [comment]
Hmm, interesting. Do you know what the intented restore mechanism was back in 2.0.0? Did we just access them from that location? I see that we expected the package to get to the `packages` dir: https://github.com/dotnet/corefx/blob/583aa597cec05337f07319056aaa648509f524c1/external/test-runtime/XUnit.Runtime.depproj#L147

I'm not sure what the best way is to deal with this now, since only Microsoft devs will be able to access that package

**Comment by trylek** [comment]
I've never become a nuget pro. @zamont may know more.

**Comment by ViktorHofer** [comment]
What are you trying to achieve @wtgodbe? I should be able to help you here.

**Comment by wtgodbe** [comment]
To re-enable multifile tests for `uapaot`. This will only work if we've restored the TestIlc Package: https://github.com/dotnet/corefx/blob/f1336fa043309e587dee5e3d59ee04249ddff97b/src/pretest.builds#L13

Which is supposed to happen here: https://github.com/dotnet/corefx/blob/a619534cac9b7d000f0899248656d232cb44f61f/external/test-runtime/XUnit.Runtime.depproj#L170-L191

But I'm not sure how we used to access the TestILC package(s). Looks like we moved them to the `packages` dir at some point, but I haven't been able to find where/how we used to do that in 2.0.0.

**Comment by ViktorHofer** [comment]
We restore ILC as part of the internal tools: https://github.com/dotnet/corefx/blob/master/eng/InternalTools.props#L10-L14

Currently we only do that for official builds. The paths in the target are still correct.

**Comment by joshfree** [comment]
@wtgodbe the landscape and infra here has likely changed a lot since this bug was filed.  It would be worth chatting with @tommcdon @zamont @nattress first before spending time on work here.

**Comment by trylek** [comment]
I'm also curious why this only affects multifile tests - I'd assume we need to pull down a reasonable version of TestILC no matter whether we build against the shared library or not.

**Comment by trylek** [comment]
In other words, I dare claim that most anything that builds in single file mode with a given TestILC drop should build in multi-file (/useSharedAssemblies) mode. Tons of work have been spent in the last 3-4 years to maintain this invariant. If there seems to be a principal problem where tests don't build in the /useSharedAssemblies mode, that's most likely an infra deficiency and I'll gladly take this offline with Will or anyone else - I was the primary implementor of the shared assembly build and use mode and of half of other ILC changes, I can certainly help figure out why a particular test build mode doesn't work as it should.

**Comment by wtgodbe** [comment]
I mostly just know what's described in this issue, and what I posted above from my brief investigation. It seems that we only use TestILC on official builds, and never do any multi-file testing - this issue just refers to turning it back on for `uapaot` (which seems to be the only Framework we ever did multi-file testing for). I'm happy to chat offline about what it means to fix this tomorrow

**Comment by wtgodbe** [comment]
So we only restore the ILC package during official builds. Should we only enable multi-file tests during those official builds? @ViktorHofer looks like we restore from `https://devdiv.pkgs.visualstudio.com/`, so probably we shouldn't ever have these run by default

**Comment by ViktorHofer** [comment]
> looks like we restore from `https://devdiv.pkgs.visualstudio.com/`

The plan is that we restore from dnceng in the future so that we don't need a secret anymore to access devdiv and enable these runs in public CI again. 

> Should we only enable multi-file tests during those official builds?

For now yes, until we bring the public jobs back in CI.

**Comment by wtgodbe** [comment]
I tried downloading the packages locally & re-enabling the multifile tests, got the following:

> MSBUILD : error : Shared assemblies 'System.Collections - System.Collections.Concurrent - System.Diagnostics.Tools - System.Diagnostics.Tracing - System.Linq - System.Linq.Expressions - System.Memory - System.ObjectModel - System.Private.CoreLib - System.Private.Interop - System.Private.Reflection.Core - System.Private.Reflection.Execution - System.Private.Reflection.Metadata - System.Private.StackTraceMetadata - System.Private.TypeLoader - System.Private.Uri - System.Reflection.TypeExtensions - System.Runtime - System.Runtime.Extensions - System.Security.Principal - System.Text.RegularExpressions - System.Threading' don't form a closure because they reference other assemblies: 'System.Buffers'. [E:\code\corefx\src\pretest.builds]
MSBUILD : error : Shared assemblies 'System.Collections - System.Collections.Concurrent - System.Diagnostics.Tools - System.Diagnostics.Tracing - System.Linq - System.Linq.Expressions - System.Memory - System.ObjectModel - System.Private.CoreLib - System.Private.Interop - System.Private.Reflection.Core - System.Private.Reflection.Execution - System.Private.Reflection.Metadata - System.Private.StackTraceMetadata - System.Private.TypeLoader - System.Private.Uri - System.Reflection.TypeExtensions - System.Runtime - System.Runtime.Extensions - System.Security.Principal - System.Text.RegularExpressions - System.Threading' don't form a closure because they reference other assemblies: 'System.Buffers'. [E:\code\corefx\src\pretest.builds]
E:\code\corefx\src\pretest.builds(15,5): error MSB3073: The command "E:\code\corefx\artifacts\bin\testhost/uapaot-Windows_NT-Release-x64/TestILC\ilc.exe -buildSharedAssemblies -buildType ret -frameworkPath "E:\code\corefx\artifacts\bin\testhost/uapaot-Windows_NT-Release-x64/ILCInputFolder"" exited with code 1201.

**Comment by trylek** [comment]
That is … sad. Apparently ongoing refactorings in CoreFX require a [most likely targeted patching-incompatible] change to the shared library definition. This is a complex situation that will require coordination with the ProjectN side. At some point I have no competence to choose (@tommcdon and @HollyAM should be the primary contacts for making the call) we should ingest the new CoreFX; this will break shared library build on our side and we'll follow up by fixing ndp\pntoolchain\VSIntegration\SharedLibraryXML\SharedLibrary.AssemblyList.txt as appropriate. @MichalStrehovsky should also have a say in this as he recently dealt with some CoreFX shuffles in ILC. Also FYI @sergiy-k and @vitek-karas.

**Comment by wtgodbe** [comment]
Sounds like re-enabling the multifile tests should be put on hold into that issue is resolved, then. Should we re-target this issue to track fixing what you've described? That isn't something I have much context on & is probably better tackled by the people you mentioned.

**Comment by trylek** [comment]
Yes, I believe your experiment uncovered that today CoreFX is out of sync with today ProjectN w.r.t. shared library build. This is something like a "cross-repo cascading build break" which needs fixing in the ProjectN closed source code before CoreFX UAPAOT multifile testing can be re-established.

**Comment by MichalStrehovsky** [comment]
> That is … sad. Apparently ongoing refactorings in CoreFX require a [most likely targeted patching-incompatible] change to the shared library definition

Targeted patching is not a concern for what's in the mainline branch. We do TP compatible releases strictly in the release/servicing branches and keep the mainline branch free-for-all.

You're right that we'll be able to flip this back to multifile testing when we flip Project N to consume 3.0 version of CoreFX. We typically do that when we have the RTM build of CoreFX. Until then, testing single file should be enough to ensure we won't get an unexpected dumpster fire in N at the time of the switch.

**Comment by HollyAM** [comment]
Just to make sure that we are all on the same page: I don't believe we have a customer scenario/request to add CoreFx 3.0 features to Project N/.NET UWP.... Do we? I believe we don't yet have a reason for Project N (.NET UWP) to support .NET Standard 2.1 and, if so, won't be until after it ships in .NET Core. Project N doesn't currently have resources allocated to add .NET Core 3.0 features, and instead is continuing to focus on the quality of existing customers using .NET Standard 2.0 (and the 2.1 version of CoreFx).

**Comment by ViktorHofer** [comment]
Closing as there is no immediate action necessary. We will probably go a different path with UWP forward.
